# Cleaning & Aggregation of the Data

Investigating the relationship between demographics and performance in high schools is an interesting angle to take to discover the efficacy of standardised tests.

Data on New York City high schools is publicly available through [NYCOpenData](https://opendata.cityofnewyork.us/), as well as the demographics of each high school.

Several data files were sourced:
- *ap_2010.csv* - Advanced Placement exam reults for each high school (passing an optional AP exam can earn a student college credit in that subject)
- *class_size.csv* - Information on class size for each school
- *demographics.csv* - Information on school demographics
- *graduation.csv* - Percentage of students who graduated for each school
- *sat_results.csv* - SAT scores by school
- *high_school_directory.csv* - Directory of NY high schools
- *survey_ny.txt* - Data on school surveys from all schools
- *survey_d75.txt* - Data on school surveys from NY District 75

In this notebook the data cleaning and aggregation process is outlined so that analysis can take place.

### Reading in and Cleaning the Data

First I will read in the data files into Pandas dataframes and quickly explore them to get a sense of the data.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
# Create a list of data files to loop through
data_files = [
    'ap_test.csv',
    'class_size.csv',
    'demographics.csv',
    'graduation.csv',
    'sat_results.csv',
    'high_school_directory.csv'
]

# Create empty dict for dataframes
data = {}

# Read csv files into Pandas dataframe
for file in data_files:
    data[file.replace('.csv', '')] = pd.read_csv(file)

In [3]:
# The survey files are txt and require different encoding
all_survey = pd.read_csv('survey_files/survey_ny.txt', delimiter='\t', encoding='windows-1252')
d75_survey = pd.read_csv('survey_files/survey_d75.txt', delimiter='\t', encoding='windows-1252')

# Concatenate the 2 survey files
survey = pd.concat([all_survey, d75_survey], axis=0)

# Create entry in data dict for survey dataframe
data['survey'] = survey

In [4]:
# Print the first few lines of each dataframe
for key in data:
    print(key)
    print(data[key].head())
    print('-----------------')

ap_test
      DBN                             SchoolName  AP Test Takers   \
0  01M448           UNIVERSITY NEIGHBORHOOD H.S.             39.0   
1  01M450                 EAST SIDE COMMUNITY HS             19.0   
2  01M515                    LOWER EASTSIDE PREP             24.0   
3  01M539         NEW EXPLORATIONS SCI,TECH,MATH            255.0   
4  02M296  High School of Hospitality Management              NaN   

   Total Exams Taken  Number of Exams with scores 3 4 or 5  
0               49.0                                  10.0  
1               21.0                                   NaN  
2               26.0                                  24.0  
3              377.0                                 191.0  
4                NaN                                   NaN  
-----------------
class_size
   CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  

Some observations:

- Some datasets have a unique DBN identifier that can be used for merging dataframes on. The DBN identifer is simply the combination of the CSD and SCHOOL CODE for each high school.
- *class_size* doesn't have a DBN column, however it does have CSD and SCHOOL CODE columns which we can use to create the DBN identifier.
- There are over 2000 columns in the survey dataframe, nearly all of which we don't need.

In [5]:
# Rename dbn columns in high_school_directory and survey to DBN for merging
data['high_school_directory']['DBN'] = data['high_school_directory']['dbn']
data['survey']['DBN'] = data['survey']['dbn']

In [6]:
# columns to keep in the survey dataframe
# these cols give us data on parent, teacher and student satisfaction about school safety and academic performance
cols_to_keep = [
    'DBN',
    'rr_s',
    'rr_t',
    'rr_p',
    'N_s',
    'N_t',
    'saf_p_11',
    'saf_t_11',
    'com_t_11',
    'eng_t_11',
    'aca_t_11',
    'saf_s_11',
    'com_s_11',
    'eng_s_11',
    'aca_s_11',
    'saf_tot_11',
    'com_tot_11',
    'eng_tot_11',
    'aca_tot_11'
]

data['survey'] = data['survey'].loc[:,cols_to_keep]

#### Formatting the DBN column in *class_size*

The DBN identifer is a combination of the CSD and SCHOOL CODE values for each high school, the only difference being the DBN is padded.

For example below the DBN value for the *sat_results* dataframe has a leading zero before the CSD number if the CSD number is only one digit i.e. the CSD portion of the DBN identifier is always two digits.

To format the DBN identfier for the *class_size* dataframe we will create a function that adds a leading zero if the CSD number is only one digit long.

In [7]:
data['class_size'].head()

CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED   
3    1       M        M015  P.S. 015 Roberto Clemente     01          CTT   
4    1       M        M015  P.S. 015 Roberto Clemente     02       GEN ED   

  CORE SUBJECT (MS CORE and 9-12 ONLY) CORE COURSE (MS CORE and 9-12 ONLY)  \
0                                    -                                   -   
1                                    -                                   -   
2                                    -                                   -   
3                                    -                                   -   
4                                    -                                   -   

  SERVICE CATEGORY(K-9* ONLY)  NUMBER OF STUDENTS / SEATS FILLED  \
0                           -                               19.0   
1                           -                               21.0   
2                           -                               17.0   
3                           -                               17.0   
4                           -                               15.0   

   NUMBER OF SECTIONS  AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  \
0                 1.0                19.0                    19.0   
1                 1.0                21.0                    21.0   
2                 1.0                17.0                    17.0   
3                 1.0                17.0                    17.0   
4                 1.0                15.0                    15.0   

   SIZE OF LARGEST CLASS DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO  
0                   19.0         ATS                             NaN  
1                   21.0         ATS                             NaN  
2                   17.0         ATS                             NaN  
3                   17.0         ATS                             NaN  
4                   15.0         ATS                             NaN

In [8]:
data['sat_results'].head()

DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404   
1                     91                             383                 423   
2                     70                             377                 402   
3                      7                             414                 401   
4                     44                             390                 433   

  SAT Writing Avg. Score  
0                    363  
1                    366  
2                    370  
3                    359  
4                    384

In [9]:
def d_check(num):
    """Function to create padded CSD numbers
    Args:
        num (int): CSD integer
    Returns:
        str: Padded CSD string
    """
    # Convert CSD integer to string
    num_str = str(num)
    
    # If CSD one digit, then add leading zero
    if len(num_str) < 2:
        return num_str.zfill(2)
    else:
        return num_str

# Apply d_check to class_size dataframe
data['class_size']['padded_csd'] = data['class_size']['CSD'].apply(d_check)
data['class_size']['DBN'] = data['class_size']['padded_csd'] + data['class_size']['SCHOOL CODE']

In [10]:
# Check it worked
data['class_size']['DBN'].head()

0    01M015
1    01M015
2    01M015
3    01M015
4    01M015
Name: DBN, dtype: object

#### Creating a single SAT score column to make analysis easier

It would be useful to have a column which totals up all of the SAT scores for different sections of the exam (math, critical reading and writing). This will make analysis of the correlations much easier.

In [11]:
# List of math, critical reading and writing SAT scores
cols = [
    'SAT Math Avg. Score',
    'SAT Critical Reading Avg. Score',
    'SAT Writing Avg. Score'
]

# Convert string to numeric
for col in cols:
    data['sat_results'][col] = pd.to_numeric(data['sat_results'][col], errors='coerce')

# Add cols together to get total sat score
data['sat_results']['sat_score'] = data['sat_results'][cols[0]] + data['sat_results'][cols[1]] + data['sat_results'][cols[2]]


In [12]:
# Print the results
data['sat_results']['sat_score'].head()

0    1122.0
1    1172.0
2    1149.0
3    1174.0
4    1207.0
Name: sat_score, dtype: float64

#### Parsing longitude and latitude of each school

Parsing longitude and latitude coordinates of each school will enable any geographic patterns in the data to be found.

The long and lat coordinates are located within the *Location 1* column in *high_school_directory*.

Functions are created to extract the lat and lon coordinates of each school.

In [13]:
import re

In [14]:
def lat_extract(string):
    """Extracts the latitude coordinate of each school
    Args:
        string (str): string located within Location 1 col
    Returns:
        str: lat coordinate
    """
    # Lat and lon values are located within brackets in the Location 1 entry
    lat_long = re.findall(r'\(.+\)', string)
    
    # Split result into lat and lon
    lat_long_ = lat_long[0].split(',')
    
    # Format the lat entry
    lat = lat_long_[0].replace('(', '')
    return lat

# Apply lat_extract over location 1 col and create lat col
data['high_school_directory']['lat'] = data['high_school_directory']['Location 1'].apply(lat_extract)

In [15]:
data['high_school_directory']['lat'].head()

0    40.601989336
1    40.593593811
2    40.692133704
3    40.822303765
4    40.773670507
Name: lat, dtype: object

In [16]:
def long_extract(string):
    """Extracts the longitude coordinate of each school
    Args:
        string (str): string located within Location 1 col
    Returns:
        str: long coordinate"""
    lat_long = re.findall(r'\(.+\)', string)
    long = lat_long[0].split(',')
    long_ = long[-1].replace(')', '')
    return long_

# Apply long_extract over Location 1 col and create long col
data['high_school_directory']['lon'] = data['high_school_directory']['Location 1'].apply(long_extract)

In [17]:
# Convert lat and lon cols to numeric
data['high_school_directory']['lat'] = pd.to_numeric(data['high_school_directory']['lat'], errors='coerce')
data['high_school_directory']['lon'] = pd.to_numeric(data['high_school_directory']['lon'], errors='coerce')

In [18]:
# Check results
data['high_school_directory'].head()

dbn                                        school_name    borough  \
0  27Q260          Frederick Douglass Academy VI High School     Queens   
1  21K559        Life Academy High School for Film and Music   Brooklyn   
2  16K393     Frederick Douglass Academy IV Secondary School   Brooklyn   
3  08X305                               Pablo Neruda Academy      Bronx   
4  03M485  Fiorello H. LaGuardia High School of Music & A...  Manhattan   

  building_code  phone_number    fax_number  grade_span_min  grade_span_max  \
0          Q465  718-471-2154  718-471-2890             9.0              12   
1          K400  718-333-7750  718-333-7775             9.0              12   
2          K026  718-574-2820  718-574-2821             9.0              12   
3          X450  718-824-1682  718-824-1663             9.0              12   
4          M485  212-496-0700  212-724-5748             9.0              12   

   expgrade_span_min  expgrade_span_max  ...  \
0                NaN                NaN  ...   
1                NaN                NaN  ...   
2                NaN                NaN  ...   
3                NaN                NaN  ...   
4                NaN                NaN  ...   

                                          Location 1 Community Board  \
0  8 21 Bay 25 Street\nFar Rockaway, NY 11691\n(4...            14.0   
1  2630 Benson Avenue\nBrooklyn, NY 11214\n(40.59...            13.0   
2  1014 Lafayette Avenue\nBrooklyn, NY 11221\n(40...             3.0   
3  1980 Lafayette Avenue\nBronx, NY 10473\n(40.82...             9.0   
4  100 Amsterdam Avenue\nNew York, NY 10023\n(40....             7.0   

  Council District Census Tract        BIN           BBL  \
0             31.0     100802.0  4300730.0  4.157360e+09   
1             47.0        306.0  3186454.0  3.068830e+09   
2             36.0        291.0  3393805.0  3.016160e+09   
3             18.0         16.0  2022205.0  2.036040e+09   
4              6.0        151.0  1030341.0  1.011560e+09   

                                                 NTA     DBN        lat  \
0  Far Rockaway-Bayswater                        ...  27Q260  40.601989   
1  Gravesend                                     ...  21K559  40.593594   
2  Stuyvesant Heights                            ...  16K393  40.692134   
3  Soundview-Castle Hill-Clason Point-Harding Par...  08X305  40.822304   
4  Lincoln Square                                ...  03M485  40.773671   

         lon  
0 -73.762834  
1 -73.984729  
2 -73.931503  
3 -73.855961  
4 -73.985269  

[5 rows x 67 columns]

#### Condensing *class_size*, *demographics* and *graduation* datasets

*class_size*, *demographics* and *graduation* dataframes all have several rows for each school. We will need to condense these dataframes so that each row pertains to a unique high school.

Exploring *class_size*, it is found that some data is relevant to school years (grades) that aren't at high school. We are only interested in grades 9-12. Each school can also have multiple programmes, GEN ED is the largest category.

We will filter *class_size* for grades 09-12 and GEN ED programme to help condense the data.

In [19]:
# Unique grade values
data['class_size']['GRADE '].unique()

array(['0K', '01', '02', '03', '04', '05', '0K-09', nan, '06', '07', '08',
       'MS Core', '09-12', '09'], dtype=object)

In [20]:
# Unique programme values
data['class_size']['PROGRAM TYPE'].unique()

array(['GEN ED', 'CTT', 'SPEC ED', nan, 'G&T'], dtype=object)

In [21]:
# Proportion of programme types
data['class_size']['PROGRAM TYPE'].value_counts(normalize=True)

GEN ED     0.556704
CTT        0.285528
SPEC ED    0.139817
G&T        0.017951
Name: PROGRAM TYPE, dtype: float64

In [22]:
# filter class_size for grades 09-12 and gen ed programme type
class_size = data['class_size']
class_size = class_size[class_size['GRADE '] == '09-12']
class_size = class_size[class_size['PROGRAM TYPE'] == 'GEN ED']
class_size.head()
# the DBN values are still not unique

CSD BOROUGH SCHOOL CODE                                    SCHOOL NAME  \
225    1       M        M292  Henry Street School for International Studies   
226    1       M        M292  Henry Street School for International Studies   
227    1       M        M292  Henry Street School for International Studies   
228    1       M        M292  Henry Street School for International Studies   
229    1       M        M292  Henry Street School for International Studies   

    GRADE  PROGRAM TYPE CORE SUBJECT (MS CORE and 9-12 ONLY)  \
225  09-12       GEN ED                              ENGLISH   
226  09-12       GEN ED                              ENGLISH   
227  09-12       GEN ED                              ENGLISH   
228  09-12       GEN ED                              ENGLISH   
229  09-12       GEN ED                                 MATH   

    CORE COURSE (MS CORE and 9-12 ONLY) SERVICE CATEGORY(K-9* ONLY)  \
225                           English 9                           -   
226                          English 10                           -   
227                          English 11                           -   
228                          English 12                           -   
229                  Integrated Algebra                           -   

     NUMBER OF STUDENTS / SEATS FILLED  NUMBER OF SECTIONS  \
225                               63.0                 3.0   
226                               79.0                 3.0   
227                               38.0                 2.0   
228                               69.0                 3.0   
229                               53.0                 3.0   

     AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  SIZE OF LARGEST CLASS  \
225                21.0                    19.0                   25.0   
226                26.3                    24.0                   31.0   
227                19.0                    16.0                   22.0   
228                23.0                    13.0                   30.0   
229                17.7                    16.0                   21.0   

    DATA SOURCE  SCHOOLWIDE PUPIL-TEACHER RATIO padded_csd     DBN  
225       STARS                             NaN         01  01M292  
226       STARS                             NaN         01  01M292  
227       STARS                             NaN         01  01M292  
228       STARS                             NaN         01  01M292  
229       STARS                             NaN         01  01M292

The DBN values for each row are still not unique. This is due to the following cols:

- CORE SUBJECT (MS CORE and 9-12 ONLY)
- CORE COURSE (MS CORE and 9-12 ONLY)

These columns look to relate to different kinds of classes. For analysis we will group by DBN and average the result to get average class sizes for each high school.

In [23]:
# group by DBN identifier and compute mean
class_size = class_size.groupby('DBN').agg(np.mean)
class_size.reset_index(inplace=True)

# update class_size dict entry
data['class_size'] = class_size

# print result
data['class_size'].head()

DBN  CSD  NUMBER OF STUDENTS / SEATS FILLED  NUMBER OF SECTIONS  \
0  01M292    1                            88.0000            4.000000   
1  01M332    1                            46.0000            2.000000   
2  01M378    1                            33.0000            1.000000   
3  01M448    1                           105.6875            4.750000   
4  01M450    1                            57.6000            2.733333   

   AVERAGE CLASS SIZE  SIZE OF SMALLEST CLASS  SIZE OF LARGEST CLASS  \
0           22.564286                   18.50              26.571429   
1           22.000000                   21.00              23.500000   
2           33.000000                   33.00              33.000000   
3           22.231250                   18.25              27.062500   
4           21.200000                   19.40              22.866667   

   SCHOOLWIDE PUPIL-TEACHER RATIO  
0                             NaN  
1                             NaN  
2                             NaN  
3                             NaN  
4                             NaN

The only column in *demographics* that prevents each row being unique is the schoolyear column.

We will only select rows where the schoolyear is 20112012 as this is the most recent year of data.

In [24]:
data['demographics'].head()

DBN                       Name  schoolyear fl_percent  frl_percent  \
0  01M015  P.S. 015 ROBERTO CLEMENTE    20052006       89.4          NaN   
1  01M015  P.S. 015 ROBERTO CLEMENTE    20062007       89.4          NaN   
2  01M015  P.S. 015 ROBERTO CLEMENTE    20072008       89.4          NaN   
3  01M015  P.S. 015 ROBERTO CLEMENTE    20082009       89.4          NaN   
4  01M015  P.S. 015 ROBERTO CLEMENTE    20092010                    96.5   

   total_enrollment prek   k grade1 grade2  ... black_num black_per  \
0               281   15  36     40     33  ...        74      26.3   
1               243   15  29     39     38  ...        68      28.0   
2               261   18  43     39     36  ...        77      29.5   
3               252   17  37     44     32  ...        75      29.8   
4               208   16  40     28     32  ...        67      32.2   

  hispanic_num hispanic_per white_num white_per male_num male_per female_num  \
0          189         67.3         5       1.8    158.0     56.2      123.0   
1          153         63.0         4       1.6    140.0     57.6      103.0   
2          157         60.2         7       2.7    143.0     54.8      118.0   
3          149         59.1         7       2.8    149.0     59.1      103.0   
4          118         56.7         6       2.9    124.0     59.6       84.0   

  female_per  
0       43.8  
1       42.4  
2       45.2  
3       40.9  
4       40.4  

[5 rows x 38 columns]

In [25]:
data['demographics']['schoolyear'].unique()

array([20052006, 20062007, 20072008, 20082009, 20092010, 20102011,
       20112012], dtype=int64)

In [26]:
demographics = data['demographics']

# filter demographics to 20112012 schoolyear
demographics = demographics[demographics['schoolyear'] == 20112012]

# update demographics dict entry
data['demographics'] = demographics
data['demographics'].head()

DBN                                              Name  schoolyear  \
6   01M015  P.S. 015 ROBERTO CLEMENTE                           20112012   
13  01M019  P.S. 019 ASHER LEVY                                 20112012   
20  01M020  PS 020 ANNA SILVER                                  20112012   
27  01M034  PS 034 FRANKLIN D ROOSEVELT                         20112012   
35  01M063  PS 063 WILLIAM MCKINLEY                             20112012   

   fl_percent  frl_percent  total_enrollment prek    k grade1 grade2  ...  \
6         NaN         89.4               189   13   31     35     28  ...   
13        NaN         61.5               328   32   46     52     54  ...   
20        NaN         92.5               626   52  102    121     87  ...   
27        NaN         99.7               401   14   34     38     36  ...   
35        NaN         78.9               176   18   20     30     21  ...   

   black_num black_per hispanic_num hispanic_per white_num white_per male_num  \
6         63      33.3          109         57.7         4       2.1     97.0   
13        81      24.7          158         48.2        28       8.5    147.0   
20        55       8.8          357         57.0        16       2.6    330.0   
27        90      22.4          275         68.6         8       2.0    204.0   
35        41      23.3          110         62.5        15       8.5     97.0   

   male_per female_num female_per  
6      51.3       92.0       48.7  
13     44.8      181.0       55.2  
20     52.7      296.0       47.3  
27     50.9      197.0       49.1  
35     55.1       79.0       44.9  

[5 rows x 38 columns]

The Demographic and Cohort cols in *graduation* prevent each row being unique.

Cohort appears to refer to the year the data represents. The most recent available is 2006 so we will choose this.

Demographic refers to a specific demographic group. We will filter on Total Cohort values as we want data from the full cohort.

In [27]:
data['graduation'].head()

Demographic     DBN                            School Name    Cohort  \
0  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2003   
1  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2004   
2  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2005   
3  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL      2006   
4  Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL  2006 Aug   

   Total Cohort Total Grads - n  Total Grads - % of cohort Total Regents - n  \
0             5               s                        NaN                 s   
1            55              37                       67.3                17   
2            64              43                       67.2                27   
3            78              43                       55.1                36   
4            78              44                       56.4                37   

   Total Regents - % of cohort  Total Regents - % of grads  ...  \
0                          NaN                         NaN  ...   
1                         30.9                        45.9  ...   
2                         42.2                        62.8  ...   
3                         46.2                        83.7  ...   
4                         47.4                        84.1  ...   

  Regents w/o Advanced - n  Regents w/o Advanced - % of cohort  \
0                        s                                 NaN   
1                       17                                30.9   
2                       27                                42.2   
3                       36                                46.2   
4                       37                                47.4   

   Regents w/o Advanced - % of grads Local - n  Local - % of cohort  \
0                                NaN         s                  NaN   
1                               45.9        20                 36.4   
2                               62.8        16                 25.0   
3                               83.7         7                  9.0   
4                               84.1         7                  9.0   

   Local - % of grads Still Enrolled - n  Still Enrolled - % of cohort  \
0                 NaN                  s                           NaN   
1                54.1                 15                          27.3   
2                37.2                  9                          14.1   
3                16.3                 16                          20.5   
4                15.9                 15                          19.2   

   Dropped Out - n Dropped Out - % of cohort  
0                s                       NaN  
1                3                       5.5  
2                9                      14.1  
3               11                      14.1  
4               11                      14.1  

[5 rows x 23 columns]

In [28]:
data['graduation']['Cohort'].unique()

array(['2003', '2004', '2005', '2006', '2006 Aug', '2001', '2002'],
      dtype=object)

In [29]:
graduation = data['graduation']

# filter graduation to 2006 cohort and total cohort demographic
graduation = graduation[(graduation['Cohort'] == '2006') & (graduation['Demographic'] == 'Total Cohort')]

# update graduation dict entry
data['graduation'] = graduation
data['graduation'].head()

Demographic     DBN                            School Name Cohort  \
3   Total Cohort  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL   2006   
10  Total Cohort  01M448    UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   2006   
17  Total Cohort  01M450             EAST SIDE COMMUNITY SCHOOL   2006   
24  Total Cohort  01M509                MARTA VALLE HIGH SCHOOL   2006   
31  Total Cohort  01M515  LOWER EAST SIDE PREPARATORY HIGH SCHO   2006   

    Total Cohort Total Grads - n  Total Grads - % of cohort Total Regents - n  \
3             78              43                       55.1                36   
10           124              53                       42.7                42   
17            90              70                       77.8                67   
24            84              47                       56.0                40   
31           193             105                       54.4                91   

    Total Regents - % of cohort  Total Regents - % of grads  ...  \
3                          46.2                        83.7  ...   
10                         33.9                        79.2  ...   
17                         74.4                        95.7  ...   
24                         47.6                        85.1  ...   
31                         47.2                        86.7  ...   

   Regents w/o Advanced - n  Regents w/o Advanced - % of cohort  \
3                        36                                46.2   
10                       34                                27.4   
17                       67                                74.4   
24                       23                                27.4   
31                       22                                11.4   

    Regents w/o Advanced - % of grads Local - n  Local - % of cohort  \
3                                83.7         7                  9.0   
10                               64.2        11                  8.9   
17                               95.7         3                  3.3   
24                               48.9         7                  8.3   
31                               21.0        14                  7.3   

    Local - % of grads Still Enrolled - n  Still Enrolled - % of cohort  \
3                 16.3                 16                          20.5   
10                20.8                 46                          37.1   
17                 4.3                 15                          16.7   
24                14.9                 25                          29.8   
31                13.3                 53                          27.5   

    Dropped Out - n Dropped Out - % of cohort  
3                11                      14.1  
10               20                      16.1  
17                5                       5.6  
24                5                       6.0  
31               35                      18.1  

[5 rows x 23 columns]

Before combining several columns in *ap_test* must be converted to numeric.

In [30]:
# list of cols to convert to numeric
cols = [
    'AP Test Takers ',
    'Total Exams Taken',
    'Number of Exams with scores 3 4 or 5'
]

# loop through cols and convert to numeric
for col in cols:
    data['ap_test'][col] = pd.to_numeric(data['ap_test'][col], errors='coerce')

# check results
data['ap_test'].dtypes

DBN                                      object
SchoolName                               object
AP Test Takers                          float64
Total Exams Taken                       float64
Number of Exams with scores 3 4 or 5    float64
dtype: object

### Combining the Data

Some data sets have missing DBN values, so different merge strategies are required.

For example, the function below demonstrates that the *ap_test* dataframe has 225 missing DBN values, and the *graduation* dataframe has 89 missing DBN values..

To preserve the data, we will use a left join when merging these with *sat_results*.

In [31]:
def dbn_checker(sat, df2):
    sat_dbns = sat['DBN'].unique()
    df2_dbns = df2['DBN'].unique()
    dbn_diff = 0
    for dbn in sat_dbns:
        if dbn not in df2_dbns:
            dbn_diff += 1
    print(dbn_diff)
    
dbn_checker(data['sat_results'], data['ap_test'])
dbn_checker(data['sat_results'], data['graduation'])

225
89


In [32]:
combined = data['sat_results']

# merge ap_test and graduation dataframes on left join
combined = combined.merge(data['ap_test'], how='left', on='DBN')
combined = combined.merge(data['graduation'], how='left', on='DBN')

combined.head()

DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers  SAT Critical Reading Avg. Score  \
0                     29                            355.0   
1                     91                            383.0   
2                     70                            377.0   
3                      7                            414.0   
4                     44                            390.0   

   SAT Math Avg. Score  SAT Writing Avg. Score  sat_score  \
0                404.0                   363.0     1122.0   
1                423.0                   366.0     1172.0   
2                402.0                   370.0     1149.0   
3                401.0                   359.0     1174.0   
4                433.0                   384.0     1207.0   

                     SchoolName  AP Test Takers   Total Exams Taken  ...  \
0                           NaN              NaN                NaN  ...   
1  UNIVERSITY NEIGHBORHOOD H.S.             39.0               49.0  ...   
2        EAST SIDE COMMUNITY HS             19.0               21.0  ...   
3                           NaN              NaN                NaN  ...   
4                           NaN              NaN                NaN  ...   

   Regents w/o Advanced - n Regents w/o Advanced - % of cohort  \
0                        36                               46.2   
1                        34                               27.4   
2                        67                               74.4   
3                       NaN                                NaN   
4                        23                               27.4   

  Regents w/o Advanced - % of grads Local - n  Local - % of cohort  \
0                              83.7         7                  9.0   
1                              64.2        11                  8.9   
2                              95.7         3                  3.3   
3                               NaN       NaN                  NaN   
4                              48.9         7                  8.3   

  Local - % of grads  Still Enrolled - n Still Enrolled - % of cohort  \
0               16.3                  16                         20.5   
1               20.8                  46                         37.1   
2                4.3                  15                         16.7   
3                NaN                 NaN                          NaN   
4               14.9                  25                         29.8   

   Dropped Out - n  Dropped Out - % of cohort  
0               11                       14.1  
1               20                       16.1  
2                5                        5.6  
3              NaN                        NaN  
4                5                        6.0  

[5 rows x 33 columns]

In [33]:
# merge rest of dataframes on inner joins as less missing DBN values
combined = combined.merge(data['class_size'], how='inner', on='DBN')
combined = combined.merge(data['demographics'], how='inner', on='DBN')
combined = combined.merge(data['survey'], how='inner', on='DBN')
combined = combined.merge(data['high_school_directory'], on='DBN')

combined.head()

DBN                                        SCHOOL NAME  \
0  01M292      HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448                UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                         EAST SIDE COMMUNITY SCHOOL   
3  01M509                            MARTA VALLE HIGH SCHOOL   
4  01M539  NEW EXPLORATIONS INTO SCIENCE, TECHNOLOGY AND ...   

  Num of SAT Test Takers  SAT Critical Reading Avg. Score  \
0                     29                            355.0   
1                     91                            383.0   
2                     70                            377.0   
3                     44                            390.0   
4                    159                            522.0   

   SAT Math Avg. Score  SAT Writing Avg. Score  sat_score  \
0                404.0                   363.0     1122.0   
1                423.0                   366.0     1172.0   
2                402.0                   370.0     1149.0   
3                433.0                   384.0     1207.0   
4                574.0                   525.0     1621.0   

                       SchoolName  AP Test Takers   Total Exams Taken  ...  \
0                             NaN              NaN                NaN  ...   
1    UNIVERSITY NEIGHBORHOOD H.S.             39.0               49.0  ...   
2          EAST SIDE COMMUNITY HS             19.0               21.0  ...   
3                             NaN              NaN                NaN  ...   
4  NEW EXPLORATIONS SCI,TECH,MATH            255.0              377.0  ...   

   priority10                                         Location 1  \
0         NaN  220 Henry Street\nNew York, NY 10002\n(40.7137...   
1         NaN  200 Monroe Street\nNew York, NY 10002\n(40.712...   
2         NaN  420 East 12 Street\nNew York, NY 10009\n(40.72...   
3         NaN  145 Stanton Street\nNew York, NY 10002\n(40.72...   
4         NaN  111 Columbia Street\nNew York, NY 10002\n(40.7...   

  Community Board Council District  Census Tract        BIN           BBL  \
0             3.0              1.0         201.0  1003223.0  1.002690e+09   
1             3.0              1.0         202.0  1003214.0  1.002590e+09   
2             3.0              2.0          34.0  1005974.0  1.004390e+09   
3             3.0              1.0        3001.0  1004323.0  1.003540e+09   
4             3.0              2.0        2201.0  1004070.0  1.003350e+09   

                                                 NTA        lat        lon  
0  Lower East Side                               ...  40.713764 -73.985260  
1  Lower East Side                               ...  40.712332 -73.984797  
2  East Village                                  ...  40.729783 -73.983041  
3  Chinatown                                     ...  40.720569 -73.985673  
4  Lower East Side                               ...  40.718725 -73.979426  

[5 rows x 161 columns]

In [34]:
# fill all missing values with the mean of the corresponding col
combined = combined.fillna(combined.mean())

# fill any further missing values with zero
combined = combined.fillna(0)
combined.head()

DBN                                        SCHOOL NAME  \
0  01M292      HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448                UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                         EAST SIDE COMMUNITY SCHOOL   
3  01M509                            MARTA VALLE HIGH SCHOOL   
4  01M539  NEW EXPLORATIONS INTO SCIENCE, TECHNOLOGY AND ...   

  Num of SAT Test Takers  SAT Critical Reading Avg. Score  \
0                     29                            355.0   
1                     91                            383.0   
2                     70                            377.0   
3                     44                            390.0   
4                    159                            522.0   

   SAT Math Avg. Score  SAT Writing Avg. Score  sat_score  \
0                404.0                   363.0     1122.0   
1                423.0                   366.0     1172.0   
2                402.0                   370.0     1149.0   
3                433.0                   384.0     1207.0   
4                574.0                   525.0     1621.0   

                       SchoolName  AP Test Takers   Total Exams Taken  ...  \
0                               0       129.028846         197.038462  ...   
1    UNIVERSITY NEIGHBORHOOD H.S.        39.000000          49.000000  ...   
2          EAST SIDE COMMUNITY HS        19.000000          21.000000  ...   
3                               0       129.028846         197.038462  ...   
4  NEW EXPLORATIONS SCI,TECH,MATH       255.000000         377.000000  ...   

   priority10                                         Location 1  \
0         0.0  220 Henry Street\nNew York, NY 10002\n(40.7137...   
1         0.0  200 Monroe Street\nNew York, NY 10002\n(40.712...   
2         0.0  420 East 12 Street\nNew York, NY 10009\n(40.72...   
3         0.0  145 Stanton Street\nNew York, NY 10002\n(40.72...   
4         0.0  111 Columbia Street\nNew York, NY 10002\n(40.7...   

  Community Board Council District  Census Tract        BIN           BBL  \
0             3.0              1.0         201.0  1003223.0  1.002690e+09   
1             3.0              1.0         202.0  1003214.0  1.002590e+09   
2             3.0              2.0          34.0  1005974.0  1.004390e+09   
3             3.0              1.0        3001.0  1004323.0  1.003540e+09   
4             3.0              2.0        2201.0  1004070.0  1.003350e+09   

                                                 NTA        lat        lon  
0  Lower East Side                               ...  40.713764 -73.985260  
1  Lower East Side                               ...  40.712332 -73.984797  
2  East Village                                  ...  40.729783 -73.983041  
3  Chinatown                                     ...  40.720569 -73.985673  
4  Lower East Side                               ...  40.718725 -73.979426  

[5 rows x 161 columns]

It will also be useful to have a column of school district values for each high school.

We can do this by exracting the first two characters of the DBN identifier.

In [35]:
def extract_district(string):
    """Extracts first two characters of DBN identifier"""
    return string[0:2]

# apply extract_district over DBN col and create school_dist col
combined['school_dist'] = combined['DBN'].apply(extract_district)

combined.head()

DBN                                        SCHOOL NAME  \
0  01M292      HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448                UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                         EAST SIDE COMMUNITY SCHOOL   
3  01M509                            MARTA VALLE HIGH SCHOOL   
4  01M539  NEW EXPLORATIONS INTO SCIENCE, TECHNOLOGY AND ...   

  Num of SAT Test Takers  SAT Critical Reading Avg. Score  \
0                     29                            355.0   
1                     91                            383.0   
2                     70                            377.0   
3                     44                            390.0   
4                    159                            522.0   

   SAT Math Avg. Score  SAT Writing Avg. Score  sat_score  \
0                404.0                   363.0     1122.0   
1                423.0                   366.0     1172.0   
2                402.0                   370.0     1149.0   
3                433.0                   384.0     1207.0   
4                574.0                   525.0     1621.0   

                       SchoolName  AP Test Takers   Total Exams Taken  ...  \
0                               0       129.028846         197.038462  ...   
1    UNIVERSITY NEIGHBORHOOD H.S.        39.000000          49.000000  ...   
2          EAST SIDE COMMUNITY HS        19.000000          21.000000  ...   
3                               0       129.028846         197.038462  ...   
4  NEW EXPLORATIONS SCI,TECH,MATH       255.000000         377.000000  ...   

                                          Location 1 Community Board  \
0  220 Henry Street\nNew York, NY 10002\n(40.7137...             3.0   
1  200 Monroe Street\nNew York, NY 10002\n(40.712...             3.0   
2  420 East 12 Street\nNew York, NY 10009\n(40.72...             3.0   
3  145 Stanton Street\nNew York, NY 10002\n(40.72...             3.0   
4  111 Columbia Street\nNew York, NY 10002\n(40.7...             3.0   

  Council District Census Tract        BIN           BBL  \
0              1.0        201.0  1003223.0  1.002690e+09   
1              1.0        202.0  1003214.0  1.002590e+09   
2              2.0         34.0  1005974.0  1.004390e+09   
3              1.0       3001.0  1004323.0  1.003540e+09   
4              2.0       2201.0  1004070.0  1.003350e+09   

                                                 NTA        lat        lon  \
0  Lower East Side                               ...  40.713764 -73.985260   
1  Lower East Side                               ...  40.712332 -73.984797   
2  East Village                                  ...  40.729783 -73.983041   
3  Chinatown                                     ...  40.720569 -73.985673   
4  Lower East Side                               ...  40.718725 -73.979426   

   school_dist  
0           01  
1           01  
2           01  
3           01  
4           01  

[5 rows x 162 columns]

In [36]:
# write clean data set to csv file for analysis
combined.to_csv('clean_data.csv', index=False)